cmd/grid.go

```diff
@@ -0,0 +1,392 @@
+package cmd
+
+import (
+	"bytes"
+	"fmt"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"text/template"
+	"time"
+
+	"github.com/metrumresearchgroup/bbi/utils"
+
+	log "github.com/sirupsen/logrus"
+
+	"os"
+
+	"github.com/metrumresearchgroup/bbi/configlib"
+
+	"github.com/metrumresearchgroup/turnstile"
+	"github.com/spf13/afero"
+	"github.com/spf13/viper"
+)
+
+// gridSpec defines how to submit a model to the grid.
+type gridSpec struct {
+	// Name is a label for the type of submission (e.g., "SGE" or "Slurm") for
+	// use in log messages.
+	Name string
+	// SubmitCommand is the name of the program used to submit to the grid
+	// (e.g., "qsub" or "sbatch").  It must accept a submission script as its
+	// one and only positional argument.
+	SubmitCommand string
+	// Template is a template used to generate the grid submission script.  It
+	// is applied to a data object with the following fields:
+	//
+	//  * Command ([]string): a slice defining the program and all arguments of
+	//    the command to execute.  These values should be rendered to the final
+	//    command string.
+	//
+	//  * Config (configlib.Config): the active configuration, which is useful
+	//    for inspecting settings like Parallel and Threads.
+	//
+	//  * JobName (string): the name that should be used for the grid submission
+	//
+	//  * WorkingDirectory (string): the directory in which the command should be
+	//    executed
+	//
+	//  The template is responsible for quoting any values as needed.
+	Template string
+	// IgnoreError is a function called if submitting the script fails.  It
+	// takes the error and combined standard output and standard error as
+	// arguments.  A return value of true indicates that the error should be
+	// swallowed rather propagated as a turnstile.ConcurrentError.
+	IgnoreError func(error, string) bool
+}
+
+// gridModel is the struct used for grid operations.
+type gridModel struct {
+	Spec                 *gridSpec
+	Nonmem               *NonMemModel
+	Cancel               chan bool
+	postworkInstructions *PostExecutionHookEnvironment
+}
```

Comment:
After this point, the code in this file is coming with light modifications from what was in `sge.go`.  Reviewing the diff locally will likely be much easier to digest.

Similarly, most of the `integration/nonmem/grid.go` content is coming unchanged from `bbi_sge_test.go`.

Reply:
@shairozan Good point, I agree it makes sense to move grid.go from under `cmd/`.

Reply:
Taking a quick look, I think that restructuring would require a larger rework because `grid.go` depends on other things defined in `cmd/` (e.g., `NonMemModel` defined `cmd/nonmem.go` and `PostExecutionHookEnvironment` defined in `cmd/run.go`).  In the long run, I don't think that larger rework would be a bad thing, though for now I'm okay with introducing a `cmd/` file that doesn't actually map to a CLI subcommand.  (We could of course absorb the code into an existing file like `cmd/run.go`, though in my view that's not an improvement organization-wise.)